### ISSUE TITLE
Scrape the entire Internet Archive in the future

### ISSUE THREAD

**Issue by upintheairsheep** [CatsLover2006/iOSobscuraServer#7]
Hello, when it is 100% confirmed the internet archive’s DDOS fiasco has finished (1-2 weeks after everything goes back up with no downtime at all in at least 1-2 weeks), it should be a good thing to scrape the entire IPA archive collection on the site, and parse everything slowly. This will be worth it, as “giant archives” do not contain everything. This will take lots of time for obvious reasons, but there’s only about 44k items on the category. Do note that the Internet Archive has a built in antivirus, and proactively takes down malware shortly after it’s upload.

**Comment by CatsLover2006** [comment]
I'll figure out how to resolve this when the fiasco is finished.

**Comment by upintheairsheep** [comment]
I think the fiasco is finished, your biggest "competition" is https://stuffed18.github.io/ipa-archive-updated/# , which seems to have partially scraped the archive.

**Comment by upintheairsheep** [comment]
https://github.com/stuffed18/ipa-archive-updated/blob/main/data/urls.json - here is a good starting point, just get the archive's identifiers and scrape them. After that scrape is complete, then scrape the rest of the Internet Archive

**Comment by CatsLover2006** [comment]
As much as I’d like to do this, the most powerful machine I have access to recently failed on me, and the next most powerful machine stuggles with the existing list of archives. Adding to this could make the full parse take over months.

**Comment by upintheairsheep** [comment]
GitHub Actions exists, you could offload the process to Actions.

**Comment by CatsLover2006** [comment]
Actions would time out. I don’t even need to check how long the theoretical max time would be, I’d hit a bandwidth limit somewhere along the way.

**Comment by upintheairsheep** [comment]
> Actions would time out. I don’t even need to check how long the theoretical max time would be, I’d hit a bandwidth limit somewhere along the way.

The GH actions time out is ~6 hours. The script would need to stop at 5h, 45min or just 5 hours. GitHub Actions does not have a bandwidth limit, learned while using tubeup on actions, and the documentation does not mention anything about a limit.

**Comment by upintheairsheep** [comment]
You should make a GitHub action to automate the scraping of the collection. In the future, other big scrapes can be automated this way as well

**Comment by CatsLover2006** [comment]
Archive.org has a bandwidth limit, and as a result the GitHub action would time out before it finished iOS obscura alone.

**Comment by CatsLover2006** [comment]
Closing as not planned at the moment (and potentially out of scope)

**Comment by upintheairsheep** [comment]
> Archive.org has a bandwidth limit, and as a result the GitHub action would time out before it finished iOS obscura alone.

The internet archive's rate limit is a generous 15 requests per minute, with a 5 minute block for offenders who exceed it, this seems doable. Doing the math, about 4500-5175 requests can be made, which is doable for GitHub actions. The Actions ToS does not forbid this type of use, they only forbid the following

**Comment by upintheairsheep** [comment]
> Cryptomining;





> Disrupting, gaining, or attempting to gain unauthorized access to, any service, device, data, account, or network (other than those authorized by the [GitHub Bug Bounty program](https://bounty.github.com/));




> `Any` activity that places a burden on our servers, where that burden is disproportionate to the benefits provided to users (for example, don't use Actions as a content delivery network or as part of a serverless application, but a low benefit Action could be ok if it’s also low burden); or



Since this is basically a temporary burden (medium-low burden, nothing compared to cracking zip files for instance) with a high benefit, but this won't constantly run after the main scrape has finished, maybe only once a month but that's going to be nothing.


> If using GitHub-hosted runners, any other activity unrelated to the **production**, testing, deployment, or publication of the software project associated with the repository where GitHub Actions are used.


This is related to the production of the project.

**Comment by CatsLover2006** [comment]
It’s not a rate limit. It’s a bandwidth limit. I have to crunch through multiple terabytes of data for just the existing set of sources. I would not be surprised if this spiked to almost a petabyte with the entire archive.

**Comment by upintheairsheep** [comment]
There is no bandwidth limit on Actions.

**Comment by CatsLover2006** [comment]
The bandwidth limit is on archive.org’s end. No matter how high the download speed is for GitHub actions, it won’t be fast enough with the download speeds archive.org provides.

**Comment by upintheairsheep** [comment]
oooh, that type of bandwidth, yeah the archive is infamous for its slow
download and upload speeds, but if it’s so bad that self-hosting is better
due to the limited time per activation, don’t consider actions.
